Add some useful Ruby functions

mckern opened the pull request:
Here's some general purpose useful Ruby functions for folks who might want to spice their scripts up a little bit!

mckern committed:
Add some useful Ruby functions

Here's some general purpose useful Ruby functions for folks who might
want to spice their scripts up a little bit!

doofus-gluteus commented on useful.rb:
```diff
@@ -0,0 +1,29 @@
+def jruby?; RUBY_ENGINE == "jruby"; end
```
One line functions are stupid.

doofus-gluteus commented on useful.rb:
```diff
@@ -0,0 +1,29 @@
+def jruby?; RUBY_ENGINE == "jruby"; end
+
+def mri?; RUBY_ENGINE == "ruby"; end
```
Another one line function ☝️ ; don't do this

doofus-gluteus commented on useful.rb:
```diff
@@ -0,0 +1,29 @@
+  Gem::Version.new(RUBY_VERSION) > Gem::Version.new("2.0.0")
+end
+
+def modern?; !legacy?; end
```
One line function. Dumb.

mckern committed:
Add a string character deduplication method!

Neat! Academically rigorous!

doofus-gluteus commented on useful.rb:
```diff
@@ -0,0 +1,42 @@
+    end
+  else
+    return false
+  end
```
![2uhf2p](https://user-images.githubusercontent.com/47807687/53298082-28018880-37dd-11e9-9d11-0daf342f96c0.jpg)

Commit:
Create `.rubocop.yaml`

If it's important enough to nitpick, then it's important enough to write linter rules!

Commit:
Update .circleci/config.yml for Ruby listing

Now that we've got some linter rules, let's apply them!

diligent-amicus commented on useful.rb:
```diff
@@ -0,0 +1,42 @@
+
+def two_one?
+  Gem::Version.new(RUBY_VERSION) <= Gem::Version.new("2.1.0")
+end
```
It looks like Rubocop doesn't like one-line method definitions (`jruby?`, `mri?`, `modern?`). Breaking these up over three lines should make the linter happy -- you can run `rubocop --auto-correct` to make these changes automatically.

diligent-amicus commented on useful.rb:
```diff
@@ -0,0 +1,42 @@
+  else
+    return false
+  end
+end
```
This method works as written, but Rubocop suggested some nitpick fixes, and using a [Guard clause](https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Style/GuardClause) instead of the `if`/`else`; I agree with that suggestion. Changing this method to use a Guard clause would look like this:

```ruby
def dedupe(str)
  return if str.empty?

  str.split("").each_with_index do |_char, i|
    if str[i] == str[i + 1]
      str.slice!(i + 1)
      return dedupe(str)
    end
  end

  str
end

>> dedupe "abbaaaaccc"
=> "abac"
```
